### PR TITLE
Added script for named entity recognition experiments

### PR DESCRIPTION
**Pull request by kenjewu**
## Description ##
This module will contain several experiments based on deep learning for named entity recognition. Currently, the model of CharCNN+BiLSTM+CRF has been added, which was proposed by [Xuezhe Ma et al. 2016](https://arxiv.org/pdf/1603.01354.pdf).

## Implemented
1. Data preprocessing.
2. Linear chain conditional random fields.
3. CharCNN+BiLSTM+CRF model.
4. Model training and evaluation with conll2003evaluate.sh.
5. Gradient clip and learning rate decay.

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
Added CharCNN+BiLSTM+CRF model to scripts/named_entity_recognition.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-466/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/6/index.html

File: scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+class CNN_BILSTM_CRF(nn.Block):
+    def __init__(self, nchars, nchar_embed, nwords, nword_embed, nfilters, kernel_size,
+                 nhiddens, nlayers, ntag_space, emb_drop_prob, rnn_drop_prob, out_drop_prob,
+                 tag2idx, **kwargs):
+        super(CNN_BILSTM_CRF, self).__init__(**kwargs)
+        with self.name_scope():
+            self.char_embedding_layer = nn.Embedding(nchars, nchar_embed)
+            self.word_embedding_layer = nn.Embedding(nwords, nword_embed)
+
+            # tanh activation or other?
```
Comment (bikestra) [inline scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py]: Why don't we reuse `convolutional_encoder` instead of writing our own character embedding? https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/model/convolutional_encoder.py

File: scripts/named_entity_recognition/code/train_helper.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        line = fin.readline()
+        fields = line.split(";")
+        acc = float(fields[0].split(":")[1].strip()[:-1])
+        precision = float(fields[1].split(":")[1].strip()[:-1])
+        recall = float(fields[2].split(":")[1].strip()[:-1])
+        f1 = float(fields[3].split(":")[1].strip())
+    return acc, precision, recall, f1
+
+
+def clip_gradients(model, clip, ctx):
```
Comment (bikestra) [inline scripts/named_entity_recognition/code/train_helper.py]: Why not use https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.utils.clip_global_norm ?

File: scripts/named_entity_recognition/index.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+		$ cd code 
+
+		$ python3 train_model.py --train "../data/eng_train.txt" --valid "../data/eng_testa.txt" --test "../data/eng_testb.txt" --wvp "../data/word_vocab.pkl" --cvp "../data/char_vocab.pkl" --tvp "../data/tag_vocab.pkl" --embedding glove --clpw 12 --nce 30 --nwe 100 --nf 30 --ks 3 --nhiddens 256 --nlayers 1 --nts 128 --edp 0.33 --odp 0.33 --rdp 0.33 0.5 --nepochs 200 --lr 0.01 -bc 16 --lds 1 --ldr 0.05 --op_name sgd --lp "../data/eval_files/logs.log"
+
+Report
+""""""""""
+The original paper used the conll2003 data to train and validate the model. The program was also 
+trained and verified on a similar conll2003 data (unofficial version). Because of the conll2003 data 
+license, the data used in this experiment is slightly different from the official data. The main
```
Comment (bikestra) [inline scripts/named_entity_recognition/index.rst]: Can we run an experiment on the official data and include metrics as well as the command, so that users shall know how we compare against others?

File: scripts/named_entity_recognition/code/train_model.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    parser.add_argument('--train', dest='train_data_path', action='store')
+    parser.add_argument('--valid', dest='valid_data_path', action='store')
+    parser.add_argument('--test', dest='test_data_path', action='store')
+    parser.add_argument('--wvp', dest='word_vocab_path', action='store')
+    parser.add_argument('--cvp', dest='char_vocab_path', action='store')
+    parser.add_argument('--tvp', dest='tag_vocab_path', action='store')
+
+    parser.add_argument('--embedding', choices=['glove', 'senna', 'random'],
+                        help='Embedding for words', required=True)
+    parser.add_argument('--clpw', type=int, dest='char_len_per_word', default=12, action='store')
```
Comment (bikestra) [inline scripts/named_entity_recognition/code/train_model.py]: I don't like the idea of using abbreviations in command line options, because that makes the shell script not readable without looking at the actual python code. Are we doing this in other scripts in the repo? I would suggest just using `--char-len-per-word` without `dest` and `store`.

File: scripts/named_entity_recognition/index.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		$ ./run_train.sh
+
+2. Train and validate the model on the specified data after configuring the custom parameters with the 
+following command.
+
+	.. code-block:: console
+
+		$ cd code 
+
+		$ python3 train_model.py --train "../data/eng_train.txt" --valid "../data/eng_testa.txt" --test "../data/eng_testb.txt" --wvp "../data/word_vocab.pkl" --cvp "../data/char_vocab.pkl" --tvp "../data/tag_vocab.pkl" --embedding glove --clpw 12 --nce 30 --nwe 100 --nf 30 --ks 3 --nhiddens 256 --nlayers 1 --nts 128 --edp 0.33 --odp 0.33 --rdp 0.33 0.5 --nepochs 200 --lr 0.01 -bc 16 --lds 1 --ldr 0.05 --op_name sgd --lp "../data/eval_files/logs.log"
```
Comment (bikestra) [inline scripts/named_entity_recognition/index.rst]: How should users get these vocab files?

File: scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py (hunk 1 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py]: Yes, this is a better idea to make the code cleaner. I will update it.

File: scripts/named_entity_recognition/code/algorithms/crf.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        score = nd.zeros((batch_size,), ctx=self.ctx)
+
+        # A matrix that retrieves the start tag of a batch of sentence symbol sequences. shape：(batch_size, 1)
+        temp = nd.array([self.tag2idx[START_TAG]] * batch_size,
+                        ctx=self.ctx).reshape((batch_size, 1))
+        # concat, shape： (batch_size, seq_len+1)
+        tags = nd.concat(temp, tags, dim=1)
+
+        def update_score(data, states):
+            '''计算评分
```
[the remark this answers is not in the document]
Comment (kenjewu) [reply scripts/named_entity_recognition/code/algorithms/crf.py]: Ok, I forgot it. I will translate the comment to english!
Strictly, we should run an experiment on offical data and report our metrics. This is better and seriously. But I have no way to get the offical conll2003 data. I have searched many data about it on the internet, but all are lightly diffierent with the offical data.  It is very beautiful that we can get the standarlize data in some way.

In addition, this is a baseline experiment to evaluate the performance of ELMo in the NER system. Therefore, if we are unable to obtain official data, we can run ELMo's experiments for NER on the same data, and we can measure the improved performance of EMLO by comparing with this baseline result.
Of course, if we have official data, everything will become more elegant.

File: scripts/named_entity_recognition/code/algorithms/crf.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+        Returns:
+            score (NDArray): value of label, shape is (batch_size, )
+        '''
+
+        batch_size = feats.shape[1]
+        score = nd.zeros((batch_size,), ctx=self.ctx)
+
+        # A matrix that retrieves the start tag of a batch of sentence symbol sequences. shape：(batch_size, 1)
+        temp = nd.array([self.tag2idx[START_TAG]] * batch_size,
```
[the remark this answers is not in the document]
Comment (kenjewu) [reply scripts/named_entity_recognition/code/algorithms/crf.py]: Ok, It is right. I can update it.

File: scripts/named_entity_recognition/index.rst (hunk 3 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/index.rst]: Strictly, we should run an experiment on offical data and report our metrics. This is better and seriously. But I have no way to get the offical conll2003 data. I have searched many data about it on the internet, but all are lightly diffierent with the offical data. It is very beautiful that we can get the standarlize data in some way.

File: scripts/named_entity_recognition/code/train_model.py (hunk 4 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/code/train_model.py]: Ok, I thought about this when I was writing the code, but I don't know which option is better. Obviously, it seems that I have chosen a not so good way. I will modify him.

File: scripts/named_entity_recognition/index.rst (hunk 5 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/index.rst]: These vocab files are the output files of the program. They can be used to predict  if you need. In short, they are just paths, not data.

File: scripts/named_entity_recognition/code/train_helper.py (hunk 2 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/code/train_helper.py]: Yes, this is just my personal habit, because at the beginning there was no such interface, now I will switch to this interface in order to make the code more concise.

File: scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py (hunk 1 above)
Comment (kenjewu) [reply scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py]: I have switched to conv_encoder, but this interface has two shortcomings. First, this interface does not provide padding. Although the effect of padding may be small in one-dimensional convolution due to globalmaxpool, it may be a little different. If I want to pad data, I need to pad the input 3D data independently, but the ndarray pad currently only supports 4d and 5d. I need to change the dimension, which adds some extra unnecessary operations. Maybe adding padding to the implementation of this interface will be easier to use.
Second, I think the input of the general convolution will be the 0th dimension of the batch, but the input of this interface is not. I have seen the implementation. it has internally performed the dimensional transformation. Is this necessary? Maybe this interface will be more intuitive when the batch is in the 0th dimension.

**Comment by mli** [comment]
Job PR-466/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/11/index.html

File: scripts/named_entity_recognition/code/train_model.py (hunk 4 above)
Comment (sravanbabuiitm) [reply scripts/named_entity_recognition/code/train_model.py]: I agree with yun on this, I think it makes sense to have --char-len-per-word kind of easily understandable notations.

**Comment by mli** [comment]
Job PR-466/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/16/index.html

**Comment by mli** [comment]
Job PR-466/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/18/index.html

File: scripts/named_entity_recognition/code/train_model.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+    all_sentences = train_sentences + valid_sentences + test_sentences
+    all_tags = train_sentences_tags + valid_sentences_tags + test_sentences_tags
+    counter = nlp.data.count_tokens(list(itertools.chain.from_iterable(all_sentences)))
+    word_vocab = nlp.Vocab(counter, max_size=50000)
+
+    char_counter = nlp.data.count_tokens(list(''.join([''.join(sentence) for sentence in all_sentences])))
+    char_vocab = nlp.Vocab(char_counter)
+
+    tag_counter = nlp.data.count_tokens(list(itertools.chain.from_iterable(all_tags)))
+    tag_vocab = nlp.Vocab(tag_counter, padding_token='O')
```
Comment (bikestra) [inline scripts/named_entity_recognition/code/train_model.py]:
```suggestion
    tag_vocab = nlp.Vocab(tag_counter, padding_token='O', bos_token=None, eos_token=None, unknown_token=None)
```
We do not need BOS/EOS/unknown tokens for tags.

**Comment by mli** [comment]
Job PR-466/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/20/index.html

**Comment by mli** [comment]
Job PR-466/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/22/index.html

**Comment by mli** [comment]
Job PR-466/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-466/22/index.html

**Comment by eric-haibin-lin** [comment]
@kenjewu any update on this?
